text/3750-cfg-os-version-min.md

```diff
@@ -0,0 +1,130 @@
+For example, when compiling for `x86_64-pc-windows-msvc` and linking with the standard library, my binary has implicitly taken a dependency on a set of APIs that Windows exposes for certain functionality.
+If I try to run my binary on older systems that do not have those APIs, then my binary will fail to run.
+When compiling for a certain target, you are therefore declaring a dependency on a minimum target API version that you rely on for your binary to run.
+
+Each standard library target uses a sensible minimum API version. for `x86_64-pc-windows-msvc` the minimum API version is "10.0.10240" which corresponds to Windows 10's initial release.
+For `x86_64-win7-pc-windows-msvc` the minimum API version is "6.1.7600" which corresponds to Windows 7.
+However, inferring the API version from the target name isn't ideal especially as it can change over time.
+
+Instead you use the `os_version_min` predicates to specify the minimum API levels of various parts of the operating system.  For example:
+
+* `os_version_min(“windows”, <string>)` would test the [minimum build version](https://gaijin.at/en/infos/windows-version-numbers) of Windows.
+* `os_version_min(“libc”, <string>)` would test the version of libc.
+* `os_version_min(“kernel”, <string>)` would test the version of the kernel.
+
+Let’s use `os_version_min(“windows”, …)` as an example.  It should be clear how this example would be extended to the other `cfg` predicates. The predicate allows you to conditionally compile code based on the set minimum API version. For example an implementation of mutex locking on Windows might look like this:
+
+```rust
+pub unsafe fn unlock(&self) {
+    *self.held.get() = false;
+    if cfg!(os_version_min(“windows”, "6.0.6000") { // API version greater than Vista
+        c::ReleaseSRWLockExclusive(raw(self)) // Use the optimized ReleaseSRWLockExclusive routine
+    } else {
+        (*self.remutex()).unlock()  // Fall back to an alternative that works on older Windows versions
+    }
+}
+```
+
+For targets where `os_version_min(“windows”, …)` does not make sense (i.e., non-Windows targets), the `cfg` predicate will return `false` and emit a warning saying that the particular `cfg` predicate is not supported on that target. Therefore, it's important to pair `os_version_min(“windows”, …)` with a `cfg(windows)` using the existing mechanisms for combining `cfg` predicates.
+
+The above example works exactly the same way with the other platform API `cfg` predicates just with different values and different target support.
+
+These predicates do not assume any semantic versioning information. The specified predicates are simply listed in order. The only semantics that are assumed is that code compiled with the `cfg` predicates works for all versions greater than or equal to that version.
+
+**Note:** Here it would be important to link to documentation showing the `cfg` predicates and the different version strings that are supported.
+
+# Reference-level explanation
```

Comment:
I think we need to discuss here what happens when you link libraries compiled with different (even though we're not specifying _how_ the user, we _will_ need a mechanism for it at some point)

Specifically, it'd be nice to talk about the pre-compiled standard library, and how it effectively becomes a requirement to use `-Zbuild-std` if the user wants to enable this kind of stuff for the standard library.